// piral-cli/lib/common/interactive.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTokenInteractively = exports.promptConfirm = exports.promptSelect = void 0;
const browser_1 = require("./browser");
const info_1 = require("./info");
const log_1 = require("./log");
const external_1 = require("../external");
function promptSelect(message, values, defaultValue) {
    const questions = [
        {
            name: 'q',
            type: 'list',
            choices: values,
            message,
            default: defaultValue,
        },
    ];
    return external_1.inquirer.prompt(questions).then((answers) => answers.q);
}
exports.promptSelect = promptSelect;
function promptConfirm(message, defaultValue) {
    const questions = [
        {
            name: 'q',
            type: 'confirm',
            message,
            default: defaultValue,
        },
    ];
    return external_1.inquirer.prompt(questions).then((answers) => answers.q);
}
exports.promptConfirm = promptConfirm;
const tokenRetrievers = {};
function getTokenInteractively(url, httpsAgent) {
    if (!(url in tokenRetrievers)) {
        const logResume = (0, log_1.logSuspend)();
        tokenRetrievers[url] = external_1.axios.default
            .post(url, {
            clientId: 'piral-cli',
            clientName: 'Piral CLI',
            description: 'Authorize the Piral CLI temporarily to perform actions in your name.',
        }, {
            headers: Object.assign(Object.assign({}, info_1.standardHeaders), { 'content-type': 'application/json' }),
            httpsAgent,
        })
            .then((res) => {
            const { loginUrl, callbackUrl, expires } = res.data;
            const now = new Date();
            const then = new Date(expires);
            const diff = ~~((then.valueOf() - now.valueOf()) / (60 * 1000));
            console.log(`Use the URL below to complete the login. The link expires in ${diff} minutes (${then}).`);
            console.log('===');
            console.log(loginUrl);
            console.log('===');
            (0, browser_1.openBrowserAt)(loginUrl);
            return external_1.axios.default
                .get(callbackUrl)
                .then(({ data }) => (Object.assign({}, data)))
                .finally(logResume);
        });
    }
    return tokenRetrievers[url];
}
exports.getTokenInteractively = getTokenInteractively;
