// piral-cli-parcel/lib/parcel/bundler-run.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.runParcel = void 0;
const events_1 = require("events");
const path_1 = require("path");
function runParcel(bundler, postProcess) {
    const eventEmitter = new events_1.EventEmitter();
    const result = {
        outFile: '',
        outDir: bundler.options.outDir,
        name: '',
        hash: '',
        requireRef: undefined,
    };
    const update = async (bundle) => {
        const requireRef = await postProcess(bundle);
        const file = bundler.mainBundle.name;
        result.hash = bundler.mainBundle.entryAsset.hash;
        result.name = (0, path_1.basename)(file);
        result.outDir = (0, path_1.dirname)(file);
        result.outFile = `/${(0, path_1.basename)(file)}`;
        result.requireRef = requireRef;
    };
    bundler.on('buildStart', () => eventEmitter.emit('start'));
    bundler.on('bundled', async (bundle) => {
        await update(bundle);
        eventEmitter.emit('end', result);
    });
    return Promise.resolve({
        async bundle() {
            const bundle = await bundler.bundle();
            await update(bundle);
            return result;
        },
        onStart(cb) {
            eventEmitter.on('start', cb);
        },
        onEnd(cb) {
            eventEmitter.on('end', cb);
        },
    });
}
exports.runParcel = runParcel;
