// piral-cli-parcel/lib/parcel/dependency.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const bundler_1 = require("./bundler");
const bundler_run_1 = require("./bundler-run");
process.on('message', async (msg) => {
    try {
        if (msg.type === 'start') {
            const bundler = (0, bundler_1.setupBundler)({
                type: 'dependency',
                externals: msg.externals,
                importmap: msg.importmap,
                targetDir: msg.targetDir,
                entryModule: msg.entryModule,
                config: {
                    outFile: msg.outFile,
                    outDir: msg.outDir,
                    cacheDir: msg.cacheDir,
                    watch: false,
                    sourceMaps: msg.sourceMaps,
                    minify: msg.minify,
                    scopeHoist: false,
                    contentHash: false,
                    detailedReport: false,
                    publicUrl: './',
                    logLevel: msg.logLevel,
                },
            });
            const handler = await (0, bundler_run_1.runParcel)(bundler, (bundle) => {
                return (0, bundler_1.postProcess)(bundle, msg.name, msg.version, true, msg.importmap, true);
            });
            await handler.bundle();
            process.send({
                type: 'done',
            });
        }
    }
    catch (error) {
        process.send({
            type: 'fail',
            error: error?.message,
        });
    }
});
